File: week1/c++/week1_8.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> heights)
{
    vector<int> answer(heights.size());

    // 뒤에서 부터 앞으로 검사
    for (int i = heights.size() - 1; i >= 0; i--)
    {
        int now = heights[i];

        int receive = 0;
        for (int j = i - 1; j >= 0; j--)
        {
            if (heights[j] > now)
            {
                receive = j + 1;
                break;
            }
        }

        answer[i] = receive;
    }
    return answer;
}

File: week1/java/week1_8.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int[] solution(int[] heights) {
        int[] answer = new int[heights.length];
        Queue<Integer> q = new LinkedList<>();

        for(int i=heights.length-1; i>=0; i--){
            boolean re = false;
            for(int j=i-1; j>=0; j--){
                if(heights[i] < heights[j]){
                    q.add(j+1);
                    re = true;
                    break;
                }
            }
            if(!re) q.add(0);
        }

        for(int i=0; i<answer.length; i++){
            answer[i] = ((LinkedList<Integer>) q).removeLast();
        }

        return answer;
    }
}